lib/commands/exam.js

```diff
@@ -254,5 +370,91 @@ module.exports = TestCommand.extend({
+  _addLoadBalancingBrowserSocketEvents: function(config) {
+    const events = config.additional_browser_socket_events || {};
+
+    events['set-modules-queue'] = function (modules) {
+      // When `load-balance` option is valid we want to have one static list of modules on server side to send a module path to browsers.
+      const proto = Object.getPrototypeOf(this);
+
+      // browserModuleMapping is defined if `replay-execution` option is being used.
+      const browserModuleMapping = this.config.progOptions.browser_module_mapping;
+
+      // Updating browserModuleMapping which maps from browser ids to a list of AMD modules ran per the browser
+      // is needed when running test suite in ci.
+      if (this.config.appMode === 'ci' && browserModuleMapping && !this.module) {
+        const browserId = /_browser=\s*([0-9]*)/.exec(this.launcher.settings.test_page)[1];
+
+        if (browserId === undefined) {
+          throw new Error('Browser Id can\'t be found.');
+        }
+
+        this.moduleQueue = browserModuleMapping[browserId];
+      } else if (!proto.moduleQueue) {
+        proto.moduleQueue = modules;
+        proto.moduleMap = {};
+      }
+    }
+
+    events['get-next-module'] = function () {
+      const proto = Object.getPrototypeOf(this);
+      const replayExecution = this.config.progOptions.browser_module_mapping;
+      const moduleQueue = this.moduleQueue;
+      const moduleName = moduleQueue.shift();
+
+      if (!moduleName) {
+        this.socket.emit('module-queue-complete');
+      } else {
+        this.socket.emit('next-module-response', moduleName);
+        // keep track of the modules ran per browserId when appMode is ci and replayExecution flag is false
+        if (this.config.appMode === 'ci' && !replayExecution) {
+          const browserId = /_browser=\s*([0-9]*)/.exec(this.launcher.settings.test_page)[1];
+
+          if (browserId === undefined) {
+            throw new Error('Browser Id can\'t be found.');
+          }
+
+          proto.moduleMap[browserId] = proto.moduleMap[browserId] || [];
+          proto.moduleMap[browserId].push(moduleName);
+        }
+      }
+
+    }
+
+    events['after-tests-complete'] = function () {
+      const proto = Object.getPrototypeOf(this);
+      const fileName = `test-execution-${Date.now()}.json`;
+
+      proto.finishedBrowsers = proto.finishedBrowsers ? proto.finishedBrowsers + 1 : 1;
+
+      const browserCount = Object.keys(config.testPage).length;
+      const moduleMapJsonObj = {
+        numberOfBrowsers: browserCount,
+        executionMapping: proto.moduleMap
+      };
+
+      if (proto.finishedBrowsers == browserCount) {
+        // Clean the moduleQueue and finishedBrowser. This guarantees new test run can set the moduleQueue. Otherwise, the server will return no more test to run.
+        // TODO: Investigate not using proto
+        delete proto.moduleQueue;
+        delete proto.finishedBrowser;
+        if (this.config.appMode === 'ci') {
```

Comment:
we should also check for`!replayExecution` as well, since we won't want to write the execution file when we are rerunning tests

Reply:
I guess we don't even need to register `after-tests-complete` event when it's running for `replay-execution` because `after-tests-complete` is just to create a json file. I'm checking if it's `replay-execution` and if it is then do not register the event at all.